File: scala/scalasearch/src/main/scala/scalasearch/FileTypes.scala
package scalasearch

import org.json.simple.parser.{JSONParser, ParseException}
import org.json.simple.{JSONArray, JSONObject}

import java.io.{IOException, InputStreamReader}
import scala.collection.mutable

object FileType extends Enumeration {
  type FileType = Value
  val Unknown: FileType = Value
  val Archive: FileType = Value
  val Binary: FileType  = Value
  val Code: FileType    = Value
  val Text: FileType    = Value
  val Xml: FileType     = Value
}

object FileTypes {
  private val _fileTypesJsonPath = "/filetypes.json"
  private val _fileTypeMap = mutable.Map.empty[String, Set[String]]

  private val archive = "archive"
  private val binary = "binary"
  private val code = "code"
  private val searchable = "searchable"
  private val text = "text"
  private val unknown = "unknown"
  private val xml = "xml"

  private def fileTypeMap: Map[String, Set[String]] = {
    if (_fileTypeMap.isEmpty) {
      val fileTypesInputStream = getClass.getResourceAsStream(_fileTypesJsonPath)
      try {
        val obj = new JSONParser().parse(new InputStreamReader(fileTypesInputStream))
        val jsonObj = obj.asInstanceOf[JSONObject]
        val ftIt = jsonObj.get("filetypes").asInstanceOf[JSONArray].iterator()
        while (ftIt.hasNext) {
          val ftObj = ftIt.next().asInstanceOf[JSONObject]
          val typeName = ftObj.get("type").asInstanceOf[String]
          val exSet = mutable.Set.empty[String]
          val exIt = ftObj.get("extensions").asInstanceOf[JSONArray].iterator()
          while (exIt.hasNext) {
             exSet += exIt.next().asInstanceOf[String]
          }
           _fileTypeMap(typeName) = Set.empty[String] ++ exSet
        }
      } catch {
        case e: ParseException =>
          print(e.getMessage)
        case e: IOException =>
          print(e.getMessage)
      }

      _fileTypeMap(text) = _fileTypeMap(text) ++ _fileTypeMap(code) ++
        _fileTypeMap(xml)
      _fileTypeMap(searchable) = _fileTypeMap(text) ++ _fileTypeMap(binary) ++
        _fileTypeMap(archive)
      Map.empty[String, Set[String]] ++ _fileTypeMap
    } else {
      Map.empty[String, Set[String]] ++ _fileTypeMap
    }
  }

  def fromName(name: String): FileType.Value = {
    val lname = name.toLowerCase
    if (lname == text) {
      FileType.Text
    } else if (lname == binary) {
      FileType.Binary
    } else if (lname == archive) {
      FileType.Archive
    } else if (lname == code) {
      FileType.Code
    } else if (lname == xml) {
      FileType.Xml
    } else {
      FileType.Unknown
    }
  }

  def getFileType(fileName: String): FileType.Value = {
    if (isCodeFile(fileName)) {
      FileType.Code
    } else if (isXmlFile(fileName)) {
      FileType.Xml
    } else if (isTextFile(fileName)) {
      FileType.Text
    } else if (isBinaryFile(fileName)) {
      FileType.Binary
    } else if (isArchiveFile(fileName)) {
      FileType.Archive
    } else {
      FileType.Unknown
    }
  }

  def isArchiveFile(fileName: String): Boolean = {
    fileTypeMap(archive).contains(FileUtil.getExtension(fileName))
  }

  def isBinaryFile(fileName: String): Boolean = {
    fileTypeMap(binary).contains(FileUtil.getExtension(fileName))
  }

  def isCodeFile(fileName: String): Boolean = {
    fileTypeMap(code).contains(FileUtil.getExtension(fileName))
  }

  def isSearchableFile(fileName: String): Boolean = {
    fileTypeMap(searchable).contains(FileUtil.getExtension(fileName))
  }

  def isTextFile(fileName: String): Boolean = {
    fileTypeMap(text).contains(FileUtil.getExtension(fileName))
  }

  def isUnknownFile(fileName: String): Boolean = {
    fileTypeMap(unknown).contains(FileUtil.getExtension(fileName)) ||
      !fileTypeMap(searchable).contains(FileUtil.getExtension(fileName))
  }

  def isXmlFile(fileName: String): Boolean = {
    fileTypeMap(xml).contains(FileUtil.getExtension(fileName))
  }

  def isZipArchiveFile(sf: SearchFile): Boolean = {
    Set("zip", "jar", "war").contains(FileUtil.getExtension(sf))
  }

  def isGzArchiveFile(sf: SearchFile): Boolean = {
    Set("gz", "tgz").contains(FileUtil.getExtension(sf))
  }

  def isBz2ArchiveFile(sf: SearchFile): Boolean = {
    "bz2" == FileUtil.getExtension(sf)
  }

  def isTarArchiveFile(sf: SearchFile): Boolean = {
    "tar" == FileUtil.getExtension(sf)
  }
}

File: kotlin/ktsearch/src/main/kotlin/ktsearch/FileTypes.kt
package ktsearch

import org.json.simple.JSONArray
import org.json.simple.JSONObject
import org.json.simple.parser.JSONParser
import org.json.simple.parser.ParseException
import java.io.File
import java.io.IOException
import java.io.InputStreamReader

/**
 * @author cary on 7/24/16.
 */

enum class FileType {
    UNKNOWN,
    ARCHIVE,
    BINARY,
    CODE,
    TEXT,
    XML
}

private const val archive = "archive"
private const val code = "code"
private const val binary = "binary"
private const val searchable = "searchable"
private const val text = "text"
private const val xml = "xml"
private const val unknown = "unknown"

fun fromName(name: String) : FileType {
    when (name.trim().toLowerCase()) {
        text -> {
            return FileType.TEXT
        }
        binary -> {
            return FileType.BINARY
        }
        archive -> {
            return FileType.ARCHIVE
        }
        code -> {
            return FileType.CODE
        }
        xml -> {
            return FileType.XML
        }
        else -> {
            return FileType.UNKNOWN
        }
    }
}

private const val fileTypesJsonPath = "/filetypes.json"

class FileTypes {

    private val fileTypeMap: Map<String, Set<String>>

    init {
        fileTypeMap = getFileTypesFromJson()
    }

    private fun getFileTypesFromJson(): Map<String, Set<String>> {
        val ftMap: MutableMap<String, Set<String>> = mutableMapOf()
        val fileTypesInputStream = javaClass.getResourceAsStream(fileTypesJsonPath)

        try {
            val obj: Any = JSONParser().parse(InputStreamReader(fileTypesInputStream))
            val jsonObj = obj as JSONObject
            val filetypesArray = jsonObj["filetypes"] as JSONArray
            for (o in filetypesArray) {
                val filetypeMap = o as Map<*, *>
                val typeName = filetypeMap["type"] as String
                val extArray = filetypeMap["extensions"] as JSONArray
                val extSet: Set<String> = extArray.map { e -> e.toString() }.toSet()
                ftMap[typeName] = extSet
            }
            val allText: MutableSet<String> = mutableSetOf()
            allText.addAll(ftMap["code"]!!)
            allText.addAll(ftMap["text"]!!)
            allText.addAll(ftMap["xml"]!!)
            ftMap["text"] = allText
            val allSearchable: MutableSet<String> = mutableSetOf()
            allSearchable.addAll(ftMap["archive"]!!)
            allSearchable.addAll(ftMap["binary"]!!)
            allSearchable.addAll(ftMap["text"]!!)
            ftMap["searchable"] = allSearchable
        } catch (e: ParseException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return ftMap
    }

    fun getFileType(file: File) : FileType {
        when {
            isTextFile(file) -> {
                return FileType.TEXT
            }
            isBinaryFile(file) -> {
                return FileType.BINARY
            }
            isArchiveFile(file) -> {
                return FileType.ARCHIVE
            }
            isCodeFile(file) -> {
                return FileType.CODE
            }
            isXmlFile(file) -> {
                return FileType.XML
            }
            else -> {
                return FileType.UNKNOWN
            }
        }
    }

    fun isArchiveFile(file: File): Boolean {
        return (fileTypeMap[archive] ?: setOf()).contains(file.extension.toLowerCase())
    }

    fun isBinaryFile(file: File): Boolean {
        return (fileTypeMap[binary] ?: setOf()).contains(file.extension.toLowerCase())
    }

    fun isCodeFile(file: File): Boolean {
        return (fileTypeMap[code] ?: setOf()).contains(file.extension.toLowerCase())
    }

    fun isSearchableFile(file: File): Boolean {
        return (fileTypeMap[searchable] ?: setOf()).contains(file.extension.toLowerCase())
    }

    fun isTextFile(file: File): Boolean {
        return (fileTypeMap[text] ?: setOf()).contains(file.extension.toLowerCase())
    }

    fun isUnknownFile(file: File): Boolean {
        return (fileTypeMap[unknown] ?: setOf()).contains(file.extension.toLowerCase())
                || !isSearchableFile(file)
    }

    fun isXmlFile(file: File): Boolean {
        return (fileTypeMap[xml] ?: setOf()).contains(file.extension.toLowerCase())
    }
}
